### ISSUE TITLE
Proposal: Need ability to add directives to host elements in component declaration.

### ISSUE THREAD

**Issue by jamesmfriedman** [angular/angular#8785]
I've been digging into Angular 2 and have run into a potential road block for extending certain kinds of components.

In the the following example, I have a button component, and a directive that will apply styles based on touch events. There will be many other objects other than just the button that will inherit the exact same touch behavior. I've explored my options, and I'm at a loss:
- Directly extend a TouchClass. This seems less than ideal since typescript doesn't support multiple class inheritance, and I'd also like to expose the behavior to consumers for use in their own classes.
- Fake multiple class inheritance through an interface. This seems like a hack and requires me to redeclare a shim api on every class I'm trying to mix into. https://www.stevefenton.co.uk/2014/02/TypeScript-Mixins-Part-One/
- Create a helper function that does it through a service directly on elementRef.nativeElement in the component constructor. I really don't want to do this since it states in the docs that nativeElement will be null when running in a worker, and that capability is something I'm the most excited about.

Without getting too deep into the guts, I'd presume that the componentMetadata is available during the components compile time, and that the host property could be scanned for additional directives that could be added dynamically and compiled at the same time. This would allow you to do mixins the angular way: using composable directives to extend functionality, and doing it without breaking view projection. Short example below.

**Current behavior**
Declaring a directive in componentMetadata.host treats it as a regular attribute

**Expected/desired behavior**
The directive declared in host would be processed at compile time.

``` typescript
/**
 * App
 */
@Component({
    selector: 'app-component',
    template: '<g-btn>TEST</g-btn>',
    directives: [gBtn, gTouch]
})

export class AppComponent {
    constructor() {

    }
}

/**
 * Touch Directive
 * Will be used in lots and lots of components
 */
@Directive({
    selector: '[g-touch]',
    host: { 
        '(touchstart)': '...',
        '(touchend)': '...',
        '(touchmove)': '...',
        '(touchcancel)': '...'
    }
})

export class gTouch {
    constructor() {

    }
}

/**
 * Simple button component
 */
@Component({
    selector: 'g-btn',
    template: '<ng-content></ng-content>',
    host: {
        'role': 'button',
        // WOULD LOVE FOR THIS TO COMPILE THE DIRECTIVE!
        // right now it just adds an attribute called g-touch
        'g-touch': ' ' 
    }
})

export class gBtn {

    constructor() {

    }
}
```

Some ideas of how this could work:

``` typescript
// Option 1: just scan the host properties for directives.
// This would be my ideal, simple and understandable
@Component({
    selector: 'g-btn',
    template: '<ng-content></ng-content>',
    host: {
        'role': 'button',
        'g-touch': true // or {prop: 'foo'} or string
    }
})

// Option 2: definitely more declarative using a hostDirectives property
// more declarative, albeit more annoying to have to reimport the touch class
@Component({
    selector: 'g-btn',
    template: '<ng-content></ng-content>',
    hostDirectives: gTouch,
    host: {
        'role': 'button',
        'g-touch': true
    }
})

// Option 3: declare host directives as its own thing, still just
// use keys pointing to bool, obj, or string
@Component({
    selector: 'g-btn',
    template: '<ng-content></ng-content>',
    hostDirectives: {
        'g-touch': {someOption: someOption}
    },
    host: {
        'role': 'button',
    }
});

// Option 4: Not a huge fan of this one, but understandable if
// people want to keep one host property
@Component({
    selector: 'g-btn',
    template: '<ng-content></ng-content>',
    host: {
        'role': 'button',
        _directives: {
            'g-touch': true
        }
    }
});
```

Thanks everyone, Angular 2 is looking great!. Let me know if theres something I'm missing.

**Comment by Andy1605** [comment]
I am currently developing a large client and am therefore trying to break down all GUI related problems into reusable Angular2 directives. This always leads me to that same problem, as james pointed out perfectly.
This really has to work somehow for the sake of good modular and dynamic architecture. The touch example is only one of many scenarios where this is needed. e.g. Drag&Drop, Resize observing, etc. etc. etc.
Made another example as plunker:
[https://plnkr.co/edit/J65THEMic0yhObt1LkCu?p=info](https://plnkr.co/edit/J65THEMic0yhObt1LkCu?p=info)

Is there any chance this functionality will be added any time soon?

**Comment by jamesmfriedman** [comment]
@Andy1605 did you ever find a way around this? I kind of tabled working with NG2 because of this during the RCs. Would love to pick it back up, but this particular issue prevents me from building extensible UI patterns.

**Comment by jofran73** [comment]
I also feel that Angular is missing an essential feature here. It should be possible for a component to declare (multiple) attribute-directives for its host. Not being able to do that is a major hindrance for my project, too.
Does anybody know if this will be implemented in the future or if there are reasons why it cannot be done?

**Comment by Aaron-Pool** [comment]
I've proposed a solution to this issue (albeit for angular version 1) here: https://github.com/angular/angular.js/issues/15270.

My idea is, instead of just having the ability to add directives, the compilation framework would have a new extensibility point called "hostTransforms" (in the case of angular 1, "nodeTransforms") which would have access to the unmodified, unfiltered component declaration and the original, uncompiled component host node whenever a component is first encountered by the compiler and being prepared for insertion into the DOM. This way a developer could extend component decorators with custom properties, and then use nodeTransforms to convert those custom properties into something the angular framework is familiar with, just before compilation. Check the feature request thread for examples.

I'm more familiar with the angular source code than the angular 2 source code, so I'm not sure if the implementation process would be the same here. But since this seems to be a pretty popular request, I'd love to see it either implemented in angular 2 and backported, or implemented in angularjs and forwardported (is that a thing?).

**Comment by oddlyfunctional** [comment]
+1

**Comment by jjstreet** [comment]
I must agree, a feature that lets us add contributing attribute directives to the host would be great. I know I could use such a feature right now implementing a more "angular" like way adding drag/drop functionality to my UI components.

What about creating a new tag similar to `<ng-container>` that lets you apply them in the component's template instead of the `host` metadata property? Something like `<ng-host [attributeDirective]>` to indicate the directives are added to the host component.

**Comment by Aaron-Pool** [comment]
@jjstreet  your proposal sounds similar to `replace: true` (obviously not identical, but similar), which was deprecated a while ago. But perhaps `replace: true` was deprecated for a reason that wouldn't apply here.

@pkozlowski-opensource Could we get a response of any kind from the ng2 team on this?

**Comment by jamesmfriedman** [comment]
I'm game for any way to achieve this. I suggested the host property because it it has access to the local scope of the component, and it already adds attributes to the component itself. Directives seem like a natural extension of this behavior.

**Comment by k3nsei** [comment]
+1 this feature is required to don't repeat code in UI components

**Comment by vlodko** [comment]
+1

**Comment by Aaron-Pool** [comment]
Could we please get some sort of response from the ng2 team on this? Even if it's just to say you're not going to do it, or to say it's a good idea but not a current priority, I'd just like to hear some sort of input.

**Comment by hccampos** [comment]
I'd like to add another use case for this. It would allow ng2-mobx (https://github.com/500tech/ng2-mobx) to get rid of the wrapping component and look much cleaner.

**Comment by klinki** [comment]
I would love to have this too. Currently I need it to make custom `routerLink` directive. I'd love to reuse angular one and just provide it with parameters prepared by mi directive.

So instead of `<a [routerLink]="repeatedCodeToGetLink()">` I would have `<a [myRouterLink]> ` and it would dynamically apply `[routerLink]` with resolved parameters.

**Comment by somombo** [comment]
Really excited about the prospects of this! 

We have needed this for a while. In fact, a while back before I knew there was an open issue for it, I [asked on stack overflow](http://stackoverflow.com/questions/41618197/consolidating-two-directives-into-a-single-one) about essentially this very feature.

I have provided an elaborate example of how we'd be able to use this feature to solve https://github.com/angular/flex-layout/issues/162 which we have had open for a while. ([See the example and explanation here](https://github.com/angular/flex-layout/issues/162#issuecomment-290350270))

We are really looking forward to any feedback, I see that this issue is the **third most thumbed-up of** all open issues in this repo. Hopefully we can see this in the next release or sooner (fingers crossed)!

/cc @tbosch @IgorMinar @mhevery @jelbourn @hansl @ThomasBurleson

**Comment by somombo** [comment]
@jjstreet I think your suggestion
```html
<ng-host myDirective="foo"></ng-host> 
```
Would go well with another separate proposal that was made a while back for seperate  reasons than we are discussing here.

See https://github.com/angular/angular/issues/7297

**Comment by fadzic** [comment]
Currently I work around this by adding directive in the parent component and then add listener in the host with the @HostListener.

Parent.html
`<my-component myDirective>`

Component.ts
`@HostListener('myEvent')
    handler() {
        // do stuff
    }`

But it would me cleaner if we can add attributes directly in the host...

**Comment by somombo** [comment]
Here is how I have been dealing with this, but I really think implementing this feature from the ground would be the best solution. 

See my self answer: http://stackoverflow.com/a/41650302/2959469

**Comment by Aaron-Pool** [comment]
Just a monthly reminder that we're waiting for positive or negative comment on this from the Angular team.

**Comment by ThomasBurleson** [comment]
@tbosch  - Any public thoughts on the priority of this issue. It also impacts `@angular/flex-layout`.

**Comment by btinoco** [comment]
@fadzic can you not just add the directive to the host element by doing this...

Component.ts
    `@HostBinding('attr.myHilite') myHilite = new myHiliteDirective()`

or like this if you need parameters like ElementRef or Renderer2
    `@HostBinding('attr.myHilite') myHilite = new myHiliteDirective(this.elementRef, this.renderer)`

I also have the need to add a directive to the host element and I was redirected to these issue.   I was able to do what I needed using the code above.  I am in no way an expert using angular but this workaround seems to work so far, if someone had issues with this approach let me know. Thanks.

**Comment by hccampos** [comment]
@btinoco that doesn't work because no lifecycle methods get called. You'd have to manually wire everything up in each component that uses the directive rather than just having Angular wire it up for you.

**Comment by btinoco** [comment]
@hccampos thanks for the heads up.  I have just tried it and the `ngOnInit` of my directive did not execute (unless I use the directive in my component manually).  Any other way to make this work? Hacks?

**Comment by ThomasBurleson** [comment]
@btinoco - yes. it is a subtle but nasty issue. One that @angular/flex-layout hopes will get fixed soon. ;-)

**Comment by Dunos** [comment]
Any news from the Angular team about this? It's been 1 year since the issue was opened...

**Comment by matheo** [comment]
Finding this detailed description on this issue was super cool,
then finding no feedback from the Angular team was super uncool :(

**Comment by tbosch** [comment]
Regarding the already working solutions:

This feature request sounds very much like mixins. In fact, bullet no 2 in the description
of this feature actually matches the official
documentation of TypeScript, see [here](https://www.typescriptlang.org/docs/handbook/mixins.html).
In angular, this becomes a little worse though, as to mixin a class with `@Input()`s, you
have to redeclare them on the base class.

Anther solution that already works today would be to make the component contain a wrapper element and apply the directives there.
E.g. if the component contained a template like `<wrapper g-touch>...`

Regarding "Create a helper function that does it through a service directly on elementRef.nativeElement":
Yes, that seems like a good idea as well. I would not care about WebWorkers for now,
as they are still experimental and are missing some bigger features for production,
and almost no library out there would work on WebWorkers. 
See e.g. also our material library which accesses the DOM directly.

**Comment by tbosch** [comment]
Regarding Option 1) of the proposal:

The current semantics for host property bindings are,
that they set a property `myDir` on the underlying HTML element,
but not any directive. However, if `host` can also introduce directives, users could write the following
and would be confused why this does not update the property in the directive `myDir`:

```
@Component({
  host: {
    '[myDir]': true
  },
  template: '...'
})
class MyComp {}
```

Regarding Option 1) and Option 3):
Introducing some kind of `host` bindings between directives on the same element can:
- lead to a cycle in the data binding graph, which Angular does not support, and therefore
  lead to hard to debug errors because of stale values / "Expression has changed after checked" errors.  
- lead to additional perf overhead, compared to directives injecting each other
  and communicating directly.

Regarding Option 2) of the proposal:
- yes, having to refer to the `gTouch` class seems odd, as all other directives
  are triggered via `NgModule`s.

**Comment by tbosch** [comment]
@ThomasBurleson let's talk offline about your use case in more details...

**Comment by somombo** [comment]
@tbosch I'd like to propose another option: Introduce a native angular tag, let's call it `<ng-host>`. 

> **Note:** @mhevery proposed the introduction of an `<ng-host>` tag in https://github.com/angular/angular/issues/7546, however, even though I'm using the same tag name here, what I am proposing is separate and meant specifically to address the issue that has been raised here.

The `ng-host` tag would not be implemented as a regular directive/component class but would instead be "magic" framework code... similar to `ng-content`, `ng-container`, etc..., 
The tag would simply serve as a "pointer" to the component host in a way that is analogous to the css [`:host` selector](https://angular.io/docs/ts/latest/guide/component-styles.html#!#-host).

It avoid ambiguous scenarios, each component would only be allowed to have, at most, one `<ng-host>` block and it would have to be the root tag/node of that component's template.

Here is how one would use it:

```typescript

// Option 5: Use `<ng-host>`.. Very declarative and intuitive
@Component({
  selector: 'g-btn',
  template: `
    <!-- 
      Besides comments, having dom code inside the template but outside of a declared 
      ng-host code block would raise an error (hopefully at compile-time) .
    -->

    <ng-host role="button" g-touch> 
      <ng-content></ng-content>
    </ng-host>
  `
})

```

**Comment by somombo** [comment]
@tbosch Thank you for responding at all. We really appreciate your engagement and feedback.

**Comment by MikeMatusz** [comment]
Are everyone else's thoughts on this functionality being specific to Components, or would it also make sense if a Directive could apply a different directive to it's host? The use case I started subscribing to this issue for involved a few third-party directives that A) we wanted to isolate from our code in case we wanted to change later and B) wanted to apply some default functionality to each instance without having to duplicate the setup every time we used it.

For example, a tooltip directive, which is going to be applied on a large number of elements throughout our application, and we want to default the delay and appendToBody (it doesn't support a centralized config object).  Because it didn't support a central config object, we had to put three or four attributes everywhere we wanted to use it.  And then later, we did end up moving away from that library (started using Material tooltips) and we had to manually replace every tooltip.  Had we been able to create our own directive that "wrapped" it,  it would have been as simple as changing that directive to apply [mdTooltip] to its host instead of the other library.

**Comment by somombo** [comment]
@MikeMatusz Looks like I had that in mind as well, here is my snippet from https://github.com/angular/flex-layout/issues/162#issuecomment-290350270.
```typescript
@Directive({
  selector: 'fxLayoutFullPage',
  hostDirectives: [LayoutDirective],
  host: { 
    'fxLayout': 'column', 
    'style': 'min-height:100vh; background-color:yellow'
  }, 
}) class LayoutFullPageDirective {}
```

**Comment by jofran73** [comment]
Would it be possible to create a property decorator that instantiates a directive?
For example:
`@HostDirective(LayoutDirective) myLayoutDirective: LayoutDirective;`

This would work for Components as well as Directives, provide an instance-reference for interaction and wouldn't get lost when inheriting from the Component/Directive.
I guess it gets more complicated if you also want to provide input and event bindings.

**Comment by soynog** [comment]
Where does this stand? I'm fairly new to Angular2/4, and what I want to do is create a directive that just applies several other directives at once. So that instead of:

```
<button directiveA directiveB directiveC>BUTTON TEXT</button>
```

I can just write:

```
<button customDirectiveABC>BUTTON TEXT</button>
```

Feels like this should be easy--basic composition/DRYness. But as far as I can tell it's not possible?

**Comment by tbosch** [comment]
Also see https://github.com/angular/flex-layout/issues/76#issuecomment-310509013

**Comment by DanielCaspers** [comment]
@soynog, I feel exactly the same. I'd also like to know where this stands.

I was hoping to be able to make draggable dialogs using Angular Material and angular2-draggable (since angular/material#1206 is not yet supported) where I'd like to dynamically add a directive to the `md-dialog-container` that `MdDialog` service creates, but it seems much harder to get the behavior of the Angular 1.x compiler here for dynamic directives. 

@tbosch , @ThomasBurleson, was the offline use case you discussed related to the issues or goals Thomas raised in angular/material#1206 by chance? I'm just trying to wrap my head around the behavior changes between the 1.6.x and 2+ frameworks.

**Comment by realappie** [comment]
Are there any updates on this issue? It gained traction in the beginning but I think its not receiving any attention anymore.

**Comment by thefill** [comment]
Yup any update on that?

**Comment by darkbasic** [comment]
This is something I need so badly, hope this proposal gets his way upstream.

**Comment by avatsaev** [comment]
This would be nice, realized today, that I can't apply directives programmatically/dynamically, got sad.

**Comment by ChrisProlls** [comment]
+1
Same for me :)
I'm searching for a way to wrap multiple directives binding into a custom directive that do all I need. For example : 

``` HTML
<my-cmp [myDirective]="content"
        [isOpen]="myCondition"
        customProp2="customClass"
        customProp1="customText">
 ...
</my-cmp>
```

It would be nice if I could create a directive which wrap all that things so I can reuse them without copy/paste all the lines.

``` HTML
<my-cmp myCustomDirective>
</my-cmp>
```

And into my custom directive 

``` HTML
<ng-host [myDirective]="content"
        [isOpen]="myCondition"
        customProp2="customClass"
        customProp1="customText">
</ng-host>
```

**Comment by jackbliss99** [comment]
coming up on the second anniversary of this issue! honestly this feature would be so, so useful, it lets us create highly composable components and directives without having to create a million wrappers. just compose the component you need from the directives you have. simple, clean, effective.

**Comment by ThomasBurleson** [comment]
@IgorMinar  - Anyway we can get this feature on the radar for upcoming enhancements?

**Comment by darkbasic** [comment]
I would like to know if such a feature would be considered a bad pattern or not. Anyone?

**Comment by ThomasBurleson** [comment]
@darkbasic  - AFAIU, without this feature, a developer would need to introduce a wrapper element (on `ng-container`) simply to add parent directives to the template view and content.

**Comment by princemaple** [comment]
No, I don't think being able to have full control of your own component without having to use wrappers is a bad pattern. It's a necessity.

**Comment by k3nsei** [comment]
@bradlygreen any comment?

**Comment by somombo** [comment]
This feature is the most popular request (if not top 5) among all open issues of this repo... Across the internet we are seeing reports (backed by empirical data) of the decline of Angular as the defacto framework ... I think one of the things driving that is a sentiment that the community is not being heard; whereas vue.js and react are gaining ground and have surpassed angular. I think this is because even though they may not necessarily implement every little thing that every one wants they atleast provide ongoing feedback about the most popular requested items. It is so frustrating to wait so long and hear nothing.. not even a simple "no, we won't do it".

**Comment by mlc-mlapis** [comment]
... although I think that some opinions about Angular / Vue / React / ... are influenced by different factors ... this concrete feature would really be worthy of some form of implementation (even circumstances are a little more complicated than just a solution with a simple list of applied directives) ... so the concrete position of Angular core team would be very welcome ... 🥇

**Comment by bradlygreen** [comment]
No specific ETA, but we're working towards making this category of things much easier in the renderer in 2018.

**Comment by somombo** [comment]
Hopefully things get drastically better in 2018. We are losing

See: https://medium.com/@chriscordle/why-angular-2-4-is-too-little-too-late-ea86d7fa0bae

**Comment by avatsaev** [comment]
@somombo this article was confirmed to be bullshit long time ago

People who actually know their stuff made fun of the author and none of them took him seriously, the likes are from react, vue fanboys.

**Comment by somombo** [comment]
So the fact of the matter is that this issue here is a **very** low priority for the angular team, in fact it is the lowest possibly priority.
> **See published priority List at [AngularHQ](https://hq.angular.io/)** (look for Issue number 8785) 

This is the case despite this issue having generated alot of discussion and interest from the community as is shown by the number of comments.


If you are someone who cares about this issue and would really like to see it implemented, then instead of waiting for.. well honestly _potentially never_, perhaps you can fill out the [Official Annual Angular Survey](https://google.qualtrics.com/jfe/form/SV_9zf3lfLkJSaVlVH), and make known that you feel like this issue should be a higher priority and would appreciate seeing it fufilled sooner rather than later.

Don't forget to thank our Angular Team for all the great work they have done!

**Comment by dszidi** [comment]
I would also like to cast my vote for this feature. This has cause a lot of grief trying to work around this issue.

**Comment by IgorMinar** [comment]
@somombo please don't read too much into the priority in AngularHQ yet. the priority formula is not fully fleshed out. having said that, I think we should revisit this feature request after v6 ships. I'm afraid that we don't have the bandwidth for it sooner and working on this would conflict with an already ongoing work in the compiler/core area.

**Comment by mhevery** [comment]
@IgorMinar the Ivy work makes this more feasible. But yes past v6.

**Comment by somombo** [comment]
@IgorMinar  and @mhevery  I can not stress enough how thankful I am (and the rest of us also, i'm sure) that you have given us this concrete feedback on what your thoughts are and what needs to happen first before this issue can be properly addressed. 

It's not always clear to us laymen what's a "quick fix" and what's not. However, barring the fact that this is not a quick fix type of thing and has to be done right, I am especially appreciative that it seems you also feel that this will be a useful feature for angular to have.

We know you are both very busy and can't possibility respond like this to every single issue. 
So you have our sincere gratitude whenever you do. We are excited and looking forward to angular v6 and beyond! 

Thank you for all the great work!

**Comment by NateMay** [comment]
You can have you component class extend or implement the directive class. If your're trying to apply the directive under the hood, then it should probably just be logic in the component.

`export class gBtn extends gTouch`

**Comment by jamesmfriedman** [comment]
@NateMay, that only allows you to extend a single class. This issue is more about composition of multiple pieces of functionality using directives.

**Comment by martindevnow** [comment]
+1 for this added feature

**Comment by simeyla** [comment]
It's possible to add a directive to an `ng-container`, which won't appear in the DOM.

I had a need for this for the intersection-observer API (which raises events when elements enter/leave the viewport). I have an `intersector` directive, which has `enter()` and `leave()` events when the element becomes visible/hidden. I have certain components that need to use this API internally, but I didn't want to add an extra DIV in the template.

So what I did was the following in `component.html` :

    <ng-container intersector (enter)="weCameOnScreen()" (leave)="byeBye()">
         ... components normal template ...
    </ng-container>

Then the `intersector.directive.ts` directive constructor injects the `ElementRef`. 

        constructor(private intersectorElementRef: ElementRef) { ... }

For a normal DOM element you would just operate on `intersectorElementRef.nativeElement`, but for an `ng-container` the node is actually a comment node. So I just check to see if it's a comment, and if it is then I go up a level.

    public ngAfterViewInit(): void 
    {
        // if the directive is applied to an ng-container must go a level up
        this.domElement = (this.intersectorElementRef.nativeElement.nodeType == 8) ? this.intersectorElementRef.nativeElement.parentElement : this.intersectorElementRef.nativeElement;

       registerIntersector(this.domElement ...);

This isn't going to work for all situations, but I'm fine with this for now. I believe that in IVY compiler they may not be using comments anymore - so this may break. The important thing was I have a single directive I can use on DOM nodes or in what is effectively as a fake '@HostBinding' for the directive.

**Comment by saberone** [comment]
I was really hoping it was possible to add directives dynamically. I want to be able to encapsulate lower level directives in "higher order", more abstract, directives. I asked the following question on stack overflow, and I was wondering whether there will be a solution for this in the future : https://stackoverflow.com/questions/51608645/abstract-away-leaflet-directive-in-custom-directive

**Comment by ermcgrat** [comment]
Subscribing to this. I also need to be able to dynamically attach a directive to a component that I created with resolveComponentFactory/createComponent.

```
const componentFactory = this.componentFactoryResolver.resolveComponentFactory(componentItem.component);

const componentRef = viewContainerRef.createComponent(componentFactory);
(<DynamicComponent>componentRef.instance).data = componentItem.data;
(<DynamicComponent>componentRef.instance).cssClassList = componentItem.cssClassList;
// Add directive to new component here
// componentRef.addDirective(someDirective)
```

**Comment by micronyks** [comment]
Any update???
I ran into another use-case where I'm using **third party directive.** 
In some scenario, I need to remove/add directive on an HTML element dynamically. 
is this possible in any way or yet pending the solution?

**Comment by mlc-mlapis** [comment]
@micronyks ... actually it's not possible to add a directive dynamically. We have to wait for Ivy which should add the possibility to create such a feature.

**Comment by micronyks** [comment]
@mlc-mlapis but any plan when will IVY come ? in which version?

**Comment by mlc-mlapis** [comment]
@micronyks ... Angular 7 by schedule.

**Comment by avatsaev** [comment]
Guys let's be reasonable here, Angular Team is trying hard to work on several huge features which are highly demanded, (PWAs, SSR, Ivy, and especially Custom Elements) the last one being very high priority feature as I could understand, because a lot of big enterprises (like Microsoft) been asking for it forever, and there is a reason for that. To achieve efficient Custom Elements they need Ivy, as soon as they're done with Ivy, as @mhevery said, the engine will allow dynamic directive more easily. 

In the meanwhile, instead of keep demanding for this feature (which I also hopelessly need btw), we can help the Angular team to accelerate the process, by testing the betas, reporting bugs, helping with the docs, etc.

Let's remember that Angular Team is not even that big, it's just a dozen of people trying to make an awesome framework for everybody, but that takes time.

**Comment by mlc-mlapis** [comment]
... yep, it's necessary to be a bit patient now and wait till the moment when we can help more with Ivy ... when the compiler will be completed and some detail design docs available.

**Comment by nayfin** [comment]
I am building a form generator, using Angular Material and Flex-Layout, which takes a JSON configuration and generates a form. This feature would help me apply the flex-layout directives to the host component at runtime. I feel like one of Angular's biggest assets is the ability to generate code at runtime from a configuration, this will go a long way to make that code more versatile. Just wanted to drop in a good use case. I'm patient ;)

**Comment by NateMay** [comment]
That’s my exact use case

**Comment by nayfin** [comment]
@NateMay [here's](https://github.com/nayfin/tft-library/tree/master/projects/tft-library) my implementation if you'd like to check it out.

**Comment by anatolibarski** [comment]
> @NateMay [here's](https://github.com/nayfin/tft-library/tree/master/projects/tft-library) my implementation if you'd like to check it out.

could you please explain? I guess you mean [( https://github.com/nayfin/tft-library/blob/master/projects/tft-library/src/lib/dynamic-form/dynamic-field.directive.ts)]

**Comment by nayfin** [comment]
The `dynamic-field.directive` does the fancy stuff but there's a lot of other stuff happening as well. I just added `CONTRIBUTING.md` in the root folder, which has instructions to setup locally. Be wary of using in anything that's in production for a couple months. I am making big changes as I work toward a stable implementation.

**Comment by dubcdr** [comment]
+1

**Comment by pauldraper** [comment]
Right now, it's impossible to add a directive to the root component.

**Comment by liujingbreak** [comment]
#### By far, my workarounds are, they all have downsides.

1. `has-it`, define new a member property as that directive inside my component class, pass all needed constructor arguments to that directive when call its constructor function (e.g. ElementRef, ViewContainerRef, TemplateRef...any injectable variables it asks for ), and manually call its lifecycle callback if it has, like `ngInit()` `ngAfterViewInit()` at current component's corresponding lifecycle callback function.

2. Wrapper everything in components template within a top-level ng-template, 
`<ng-template><div targetDirective>....</div></ng-template>` render them in `ngAfterViewInit()` like:
```ts
const vf = this.viewContainerRef.createEmbeddedView(this.templateRef);
vf.detectChanges();
```
This way, Angular creates another `element` with that directive on it and my component's actual content inside it right after the original component element in DOM tree.
```html
<my-component></my-component>
<div targetDirective>....</div>
```
This way is like what `<route-outlet>` does.

There are obvious side effects

**Comment by pauldraper** [comment]
> Let's remember that Angular Team is not even that big, it's just a dozen of people trying to make an awesome framework for everybody, but that takes time.

It could be bigger by having a community of contributors.

However, the change of a contributed fix for this being accepted is very low.

So instead we're back to a dozen people.

**Comment by LaniMoo** [comment]
> Can someone confirm if this is now possible with Ivy? If so, does anyone have an example?

Since no word yet, thought I'd provide the closest thing I was able to find, which is an article from a while back about implementing mixins with Ivy: https://blog.nrwl.io/metaprogramming-higher-order-components-and-mixins-with-angular-ivy-75748fcbc310

According to the article, I think a possible solution for the original issue of this thread is to use the new feature called... "features". 

...You can imagine it's a bit of a nightmare to try to Google anything about this feature. Hoping they release some official Ivy documentation soon! :)

**Comment by nop1984** [comment]
@nayfin also building visual form designer/builder
And after few months of working to get stuck on fact I have no way to deploy directive to dynamically added div makes me crazy ....

**Comment by tsteuwer** [comment]
This feature would be overwhelmingly welcomed as I find myself always making a single `div` to attach top-level directives. Additionally, If I want any flex-layout directives, I have to make that one-off div as well and it would be nice if I can attach them directly to the host element instead of having to do this.

**Comment by stephenlautier** [comment]
Would be super cool to be able to add directives dynamically such as:

```ts
const msked = componentFactory.createDirective(MaskedInputDirective);
msked.textMaskConfig = {};
this.elementRef.directives.add(msked);
```

**Comment by BAM5** [comment]
> Additionally, If I want any flex-layout directives, I have to make that one-off div as well and it would be nice if I can attach them directly to the host element instead of having to do this.

@tsteuwer You could always just use the :host selector in your scss to apply style properties to the host element.

**Comment by pauldraper** [comment]
> Wrapper everything in components template within a top-level ng-template

A slightly slicker alternative to this is to use https://github.com/trotyl/angular-contrib and add

```
host: { ngNoHost: '' }
```

This project shims the renderer and prevents rendering the host element.

It has many of the same drawbacks of course.

**Comment by petriq** [comment]
Pity this is still open after 3 years. Directives bound to host element would really improve code reuse-ability.

**Comment by BenLune** [comment]
> > Additionally, If I want any flex-layout directives, I have to make that one-off div as well and it would be nice if I can attach them directly to the host element instead of having to do this.
> 
> @tsteuwer You could always just use the :host selector in your scss to apply style properties to the host element.
> 
> But yeah, I also would like the ability to apply directives to the host element. would be useful for making the host element scrollable and applying CdkScrollable from Angular Material CDK.

Not ideal but you can create the CdkScrollable programmatically this way :
this.scrollable = new CdkScrollable(this.elementRef, this.scrollDispatcher, this.zone);
this.scrollable.ngOnInit();

You have to destroy it manually too :
if (this.scrollable) {
      this.scrollable.ngOnDestroy();
}

**Comment by somombo** [comment]
> https://github.com/angular/angular/issues/8785#issuecomment-361004682 IgorMinar the Ivy work makes this more feasible. But yes past v6.

@mhevery Following up on your comment :point_up_2:, now that ivy has finally fully landed, can we pretty please have this feature for v10? If not, kindly update us on what other considerations may hold this back further.

**Comment by Urbanz91** [comment]
Are any changes?

**Comment by princemaple** [comment]
If this specific feature was on Angular survey https://twitter.com/angular/status/1252646001162088448?s=20, I bet it'd be the top voted entry.

**Comment by darkbasic** [comment]
There are tons of features which would be top voted, this one for sure but also using Observables for @output and many others. Unfortunately at the current pace they will never get implemented.

**Comment by somombo** [comment]
> If this specific feature was on Angular survey, I bet it'd be the top voted entry.

Great idea @princemaple!

Though not ideal, this can actually be addressed in the survey's "extra comments" section (of question 2)
Where it says:

```
"How else should we improve Angular for you?"
```

**So basically, everyone interested in seeing this feature, please just answer the survey and make it *explicitly* known, that you care *alot* about seeing "Issue #8785" implemented and resolved.**

Here is the direct link to the survey:
https://goo.gle/angular-survey-2020

May the force be with you! 🙂

**Comment by ronnyek** [comment]
I too have struggled with how to programatically add more functionality to components, and honestly I think some of the proposals here seem like the BEST ways of approaching those specific problems.

I've spoken with angular team members regarding that article 

> > Can someone confirm if this is now possible with Ivy? If so, does anyone have an example?
> 
> Since no word yet, thought I'd provide the closest thing I was able to find, which is an article from a while back about implementing mixins with Ivy: https://blog.nrwl.io/metaprogramming-higher-order-components-and-mixins-with-angular-ivy-75748fcbc310

And basically was led to the impression this is hacking with angular's internals, and it was in fact not designed for typical user consumption.

I'm not sure what if any technical reason exists that prevents us from being able to do this, but I think if we had the capabilities to do this, it would DRASTICALLY improve my day to day with angular.

**Comment by somombo** [comment]
> “We’ve dramatically increased our investment in working with the community. In the last three weeks our open issue count has decreased by over 700 issues across framework, tooling, and components. We’ve touched over 2,000 issues, and we plan to make large investments over the next few months, working with the community to do even more.” — @StephenFluin 
from [Angular 10 Announcement](https://blog.angular.io/version-10-of-angular-now-available-78960babd41)

So I guess this means we'll see this issue quashed in v11? 🤞😏

What better a way is there to "work with the community" (and appease them) than to work on adding one of their most requested features!? (this one 😉)

Listen to them!

**Comment by mhevery** [comment]
Just to set the expectations, what you are asking is not trivial amount of work and the current data structures are not really designed for this. So to support something like this would require some major engineering.

**Comment by k3nsei** [comment]
@mhevery how it's different from applying them from parent?

**Comment by mlc-mlapis** [comment]
[**@k3nsei**](https://github.com/k3nsei) It's necessary to see it from `NgModule` point of view, which is actually the key element that creates the infrastructure for all of its components.

**Comment by k3nsei** [comment]
@mlc-mlapis We have @HostBinding and @HostListener so maybe @HostDirective would be good choice for that functionality.

**Comment by mlc-mlapis** [comment]
[**@k3nsei**](https://github.com/k3nsei) It could be, but I am not sure if it's not too dynamic, so less performant than strictly static structures.

**Comment by somombo** [comment]
> "Just to set the expectations, what you are asking is not trivial amount of work and the current data structures are not really designed for this. So to support something like this would require some major engineering." — https://github.com/angular/angular/issues/8785#issuecomment-654391378

Thanks for your timely response @mhevery. 

I think I'll be speaking for community in saying that it is not at all lost on us that this will be a very big challenge. If it weren't, I am sure by now we would have made some third-party libraries that achieves this properly (somehow). [to my knowledge there isn't any]. 

Also, goes without saying but, please let us know if there are some low hanging fruit (or otherwise) that we can assist with contributing towards this. 

We sincerely thank you and value your earnest communication and hope we will continue to be a part of the dialogue on what we need vs. what is realistic/pragmatic to add.

**Comment by pauldraper** [comment]
Though my understanding is that Ivy makes this easier than before.

@mhevery 

> @IgorMinar the Ivy work makes this more feasible. But yes past v6.

**Comment by somombo** [comment]
> Though my understanding is that Ivy makes this easier than before

My new understanding is "easier" still doesn't mean "easy"

**Comment by pauldraper** [comment]
> My new understanding is "easier" still doesn't mean "easy"

Ivy: The thing you spent two years on and still doesn't address any of the most popular Angular issues.

**Comment by somombo** [comment]
> Ivy: The thing you spent two years on and still doesn't address any of the most popular Angular issues.

@pauldraper I guess our issues are not their "issues" seeing as this particular one isn't even on their radar (See Roadmap https://angular.io/guide/roadmap).

For me personally, I think its time for me to look elsewhere for a project that's not only open source but a project whose direction the community (and users) have a *real* influence on.

**Comment by m-radzikowski** [comment]
> @pauldraper I guess our issues are not their "issues" seeing as this particular one isn't even on their radar (See Roadmap https://angular.io/guide/roadmap).

@somombo I am disappointed by the fact that this issue is still open after all these years, but I cannot agree with you  The first point on the roadmap is explicitly about handling open github issues. Listing all of them on the roadmap wouldn't make much sense, would it? This issue is one of the most upvoted (or **the** most upvoted) and I hope it will be finally taken resolved.

**Comment by fxck** [comment]
> Listing all of them on the roadmap wouldn't make much sense, would it? This issue is one of the most upvoted (or the most upvoted) and I hope it will be finally taken resolved.

nope, this is just wishful thinking

**Comment by somombo** [comment]
> > @pauldraper I guess our issues are not their "issues" seeing as this particular one isn't even on their radar (See Roadmap https://angular.io/guide/roadmap).
> 
> @somombo I am disappointed by the fact that this issue is still open after all these years, but I cannot agree with you The first point on the roadmap is explicitly about handling open github issues. Listing all of them on the roadmap wouldn't make much sense, would it? This issue is one of the most upvoted (or **the** most upvoted) and I hope it will be finally taken resolved.

All the same .. I am done waiting.. this issue has litterally been a huge blocker for me. So the fact that it doesn't seem like it's even being planned for any time soon means it's time for me personally to move on. Best of luck to everyone else.

**Comment by johncrim** [comment]
I would like to see this renamed to "Support for adding directives to directives". While that name might be confusing, I think it's important that this feature work on directives, and not be limited to components. Other names for the feature might be "implied directives" or "attached directives", meaning when you use a given component or directive in a template, it pulls in the implied/attached directives on the host element.

I've wanted this many times, mainly because composition is potentially a cleaner form of re-use in Angular, compared to inheritance. Inheritance can be clunky because there's no multiple inheritance, constructor parameters have to be passed down, and some Angular features work when inherited, and others have to be "re-connected" in each leaf class.

I imagine "implied/attached directives" working like a component-local or directive-local form of directive instantiation, which results in the directive being instantiated on the host element without requiring the directive selector to exist in the template markup.

Here's an example:

```ts
@Directive({
  selector: 'app-popup',
  attachDirectives: [
    FocusAreaDirective
  ]
})
export class PopupDirective {

  // Attached directives can be injected, just like template-declared directives today
  constructor(public focusArea: FocusAreaDirective) {
  }

}
```

The `@Input()` and `@Output()` properties on the attached directive can be used in the template, and the lifecycle methods on the attached directive should be called as part of the host component lifecycle. The attached directive can also bind to the host element. Basically, it acts exactly like a template-declared directive today, but doesn't need to be declared in the template.

I think this feature would provide a significant benefit to Angular, enabling cleaner/simpler component architectures via directive composition.

Today, you have 2 choices if you want a similar form of directive re-use:

* Require that a set of related directives always be declared together in template markup; and detect and throw errors if the required directives aren't present in the constructor. There's no way to require that the required directives are declared on the same element. This is messy from a template authoring and documentation standpoint, and is not a strong API or contract due to redundancy, but is otherwise not horrible.
* Manually instantiate the attached directives within the host directive, and forward constructor parameters, @Input/@Output properties, host bindings, and lifecycle methods to the internal directives. This is a fragile mess for component authors, but is doable given a simple set of related components. It is much nicer for template authoring.

To put it another way, the absence of the feature sometimes creates an unnecessary tradeoff between clean+simple component use, or clean+simple component authoring.

**Comment by amakhrov** [comment]
Note that in a real world case your custom directive would have inputs, and you would want to transform them and pass as inputs to attached directive. Perhaps this could be done with a syntax similar to `host` attributes in Directive decorator options.

**Comment by johncrim** [comment]
@amakhrov : Good point - I excluded the inputs from my example for clarity. In most of the cases where I need this, I don't need to transform the inputs (or outputs) for the attached directives - they (ideally) act as composable units and their input (or output) values could be bound from the template using the parent directive (or component).

In cases where there are either naming conflicts or naming clarity issues (which I would try to avoid when designing directives for composition), or when inputs or outputs needs to be transformed, that could be handled pretty easily by injecting the attached directive in the parent directive, and create new input or output properties on the parent that delegate to the attached directives.

**Comment by somombo** [comment]
I stand corrected. 
This issue is now listed under the "Future" section of the official roadmap. See https://angular.io/guide/roadmap#support-adding-directives-to-host-elements

> ### Support adding directives to host elements
> A long-standing feature request is to add the ability to add directives to host elements. The feature will allow developers to augment their own components with additional behaviors without using inheritance. The project will require substantial effort in terms of the definition of APIs, semantics, and implementation.

Since I have just noticed it, I'm not sure when it was added,  but I will admit that this is great news, and a meaningful and reassuring gesture. I'll keep crossing my fingers.

Thank you to the team for putting it there!

**Comment by HyperLife1119** [comment]
Five years...

**Comment by nithintel** [comment]
Sorry for trouble very new to angular - not experienced stackoverflow i posted but they declined the question

I saw your stackoverflow answer, Please help me to sort one issue
on the Angular 8 - actually i have create stakblitz demo the problem i am facing is in the dropdown menu when i select value it has to display the selected item

It current binds to the html value and displays the list of value selected 
And there is add button when i click add button another dropdown should add it adds

but the issue is when i select the second dropdown value the first dropdown value selected also changes

I dont want that to happen the first dropdown value has to be same as selected and the second dropdown value should not affect the first dropdown value

I could not find any example in stackoverflow

Please help sharing the link of stackblitz

[](https://stackblitz.com/edit/angular-ivy-mq7xne?file=src%2Fapp%2Fcomparison-production%2Fcomparison-production.component.html)


image.png

Now when i select the celeritas the First selected value changed, the should not happen it has first value of first dropdown, and second value of second dropdown


image.png

**Comment by Airblader** [comment]
@nithintel There's 61 people watching this issue, please don't abuse it for unrelated questions. If your question on SO was closed, it happened for a reason, and you should ask it again providing all relevant information and an example to reproduce it. This issue is not the place for that.

**Comment by RobertAKARobin** [comment]
This is one of the biggest disadvantages of Angular compared to Vue, IMO.

**Comment by pauldraper** [comment]
I hate to spam the thread, but this has been on the roadmap for over a year without a single update from the Angular team.

Is this *actually* on the roadmap?

**Comment by nithin-neewee** [comment]
@pauldraper  is the angular team been separated hearing news that they are not interested in the support or the contribution to it.

**Comment by mgechev** [comment]
We just started investigation of the scope of the work. Depending on the result, we'll either follow up with an RFC that has an attached design document, or explanation what the next steps are.

**Comment by basvdvelden** [comment]
Might be time to start thinking about considering the option to ponder whether or not to finally add this feature perhaps? Seems kind of like basic functionality which has now been open for 6 years...

**Comment by crisbeto** [comment]
Sorry for the delayed response, I just reached this notification in my inbox. I've been exploring this feature and I have a [working prototype](https://github.com/crisbeto/angular/commit/7648e5c2507e216cd05719a66a2c41425b273469) that allows you to apply directives through the `hostDirectives` property on `@Component`/`@Directive`. 

That being said, since this is a fundamental change to how directives currently work, we'll have to spend some time making sure that the behavior makes sense.

**Comment by akleimrey** [comment]
I'll be frank, I have been counting myself among the "fans" of this framework. But the fact that one can't apply directives to components makes most applications of directives close to useless, as in many cases we have to break the encapsulation of components. Keeping components "dumb" remains especially challenging, once someone works with Drag'n Drop, Drawing and so on, especially since components are always wrapped in component tags! I'm currently reusing a custom directive to unwrap components to make some intermediate DOM manipulations more sensible.

**Comment by Airblader** [comment]
> especially since components are always wrapped in component tags

That's not true, components can use attribute selectors, too.

> makes most applications of directives close to useless

Neither is this, nor do I understand what it would have to do with this issue.

**Comment by chaosmonster** [comment]
> That being said, since this is a fundamental change to how directives currently work, we'll have to spend some time making sure that the behavior makes sense.

I really like the usage of this approach, it would be clear and simple. 
Would it still work to set the directive's input if the host component doesn't have the input defined? I would expect this to be possible (did not test it)

**Comment by princemaple** [comment]
Did Emma suggest that it's going to be in v15? https://twitter.com/twerske/status/1545100005425303552

**Comment by pauldraper** [comment]
Yes, she did! https://twitter.com/twerske/status/1545104861762445313

**Comment by simeyla** [comment]
I really hope this is implemented via a special 'ng' element such as `<ng-host>` in the template. 

If this *is* how it will be done I really hope at the same time we'd then be able to add host binding attributes, classes, styles and events to it.

```
<ng-host [attr.attr1]="true" 
         [class.theme-1]="true"
         [style.outline]="model.outline$ | async" 
         [intersectionObserver]="intersectionObserverOptions"
         (click)="hostClicked()">

   Normal template...

</ng-host>
```

HostBinding can just get too cumbersome right now and if a lot of re-engineering is needed for this feature I really hope we can get some other benefits like this as a side effect.

In fact I'd be more excited to be able to move `HostBinding` logic to the template than add directives to host elements. Especially if it means I can finally use async pipe for host bindings and thereby fix this issue https://github.com/angular/angular/issues/19483.

**Comment by crisbeto** [comment]
It is being implemented as a `hostDirectives` property on `@Directive` and `@Component`:

```
@Component({
  hostDirectives: [DirectiveA, {directive: DirectiveB, inputs: ['foo: fooAlias'], outputs: ['bar']}]
})
```

We decided against the `ng-host` element, because:

1. It introduces a third way of doing host bindings, on top of the host property and the `@Host*` decorators. New users may find it confusing and we’ll have to figure out how to handle conflicting information.
2. It does not allow for host directives to be applied to other directives, only to components.
3. It allows users to place elements in weird places, e.g. `ng-content` outside `ng-host` or `*ngIf` on the host.
4. Having `<ng-host>` inside the component’s template is misleading, because the change detection occurs in the parent view.

**Comment by Harpush** [comment]
> It is being implemented as a `hostDirectives` property on `@Directive` and `@Component`:
> 
> ```
> @Component({
>   hostDirectives: [DirectiveA, {directive: DirectiveB, inputs: ['foo: fooAlias'], outputs: ['bar']}]
> })
> ```
> 
> We decided against the `ng-host` element, because:
> 
> 1. It introduces a third way of doing host bindings, on top of the host property and the `@Host*` decorators. New users may find it confusing and we’ll have to figure out how to handle conflicting information.
> 2. It does not allow for host directives to be applied to other directives, only to components.
> 3. It allows users to place elements in weird places, e.g. `ng-content` outside `ng-host` or `*ngIf` on the host.
> 4. Having `<ng-host>` inside the component’s template is misleading, because the change detection occurs in the parent view.

Just curious - I got a few questions if I may:
1. inputs can be set in the component decorator and with `@Input` decorator - will it be supported here too with a new decorator?
2. Will it be type safe? Especially the inputs and outputs example.
3. Can we pass those inputs from the component or just hard coded values? And if we will be able - will it support passing data from component inputs (Delegate)?
4. Will there be support for dynamic directives placement? For example based on some runtime check. It actually has two levels - either in constructor before component initialize allowing DI based decision or fully dynamically at all times (some kind of imperative API for directive adding and removing)
5. What does it mean apply directive to other directives? I can see for example when you want to add a certain directive whenever another is added - but won't it conflict when a component is the host? If both the component and the directive applied to this component declare the same host directive - what will happen?
6. Lastly - like question 5 - applying a directive from outside on the host and from inside will result in a conflict - will it be an error? What's expected here?

**Comment by Airblader** [comment]
> Will there be support for dynamic directives placement? For example based on some runtime check.

You can't even do that in the template currently, so I'd be surprised to see it added here. I've just gotten used to writing my directives such that they can be toggled.

**Comment by crisbeto** [comment]
> inputs can be set in the component decorator and with @Input decorator - will it be supported here too with a new decorator?

> Can we pass those inputs from the component or just hard coded values? And if we will be able - will it support passing data from component inputs (Delegate)?

I think that both of these have the same answer so I'll address them here. You won't be able to bind directly from the host to the host directive. One of our earlier proposals looked something like this:

```
@Component({
  hostDirectives: [{
    directive: MyDir,
    bindings: {
      '[someInput]': 'someValue',
      '(someEvent)': 'someCallback($event)'
    }
  }]
})
class MyComp {}
```
We decided not to implement it, because it would've been prone to "changed after checked" errors and all the workarounds would've introduced a lot of complexity in the codebase and potentially performance issues. The `inputs` and `outputs` syntax in my example above determines which inputs/outputs from the host directive will be available on the host. By default none of inputs/outputs are available so that the directive author has more control over the public API. This handles the primary use case for the `bindings` syntax which was being able to alias inputs and outputs.
If you want to interact directly with the host directive instance, you can use DI since it's going to be published on the same injector as the host. E.g. you'll be able to do something like:

```
@Component({
  hostDirectives: [MyDir]
})
class MyComp {
  constructor(myDirInstance: MyDir) {}
}
```


> Will it be type safe? Especially the inputs and outputs example.
Type safe in what way? It is going to be integrated with our existing template type checking so if you do something like `<my-comp [inputFromHostDirective]="foo"></my-com>`, you'll get type checking of the `inputFromHostDirective` property.



> Will there be support for dynamic directives placement? For example based on some runtime check. It actually has two levels - either in constructor before component initialize allowing DI based decision or fully dynamically at all times (some kind of imperative API for directive adding and removing)
No, because we wouldn't be able to type check host directives that are applied dynamically.

> What does it mean apply directive to other directives? I can see for example when you want to add a certain directive whenever another is added - but won't it conflict when a component is the host? If both the component and the directive applied to this component declare the same host directive - what will happen?

It means that you'll be able to set `hostDirectives` on `@Directive` instances as well, allowing you to create chains of host directives. We thought that this would be valuable to component library authors who may need more advanced composition of directives. 

> Lastly - like question 5 - applying a directive from outside on the host and from inside will result in a conflict - will it be an error? What's expected here?
If the same directive is matched multiple times through some combination of `hostDirectives` and the template where the host is being used, you'll get an error from the framework since a lot of things will break down if a directive applies more than once to the same element. Initially we considered de-duplicating the instances during directive matching, but it would've been tricky to determine the correct execution order of the directives.

**Comment by Harpush** [comment]
@crisbeto By type safe I also meant the host directives inputs property in the component metadata. In your example you wrote `inputs: ['foo: fooAlias']` - I meant if `foo` isn't an input will it give a compile time error?

**Comment by crisbeto** [comment]
It'll get the same level of type checking as the existing `inputs` property on `@Directive` and `@Component`. I don't think that we check if the property exists since it might be inherited from a parent class, but you should get a type checking error when you try to bind to it.

**Comment by nartc** [comment]
Moved from PR

@crisbeto thank you for the response. I figured it'd be the case. In that case, should we update the PR description to call it out for Structural Directives? I guess it makes sense that this feature only targets Attribute Directive. Structural should be handled on the consuming (component's users) side of things rather than authoring (component's author).

Another question: is it possible to have TypeScript generics supported?

```ts
@Directive()
export abstract class NgtInstance<TObject> {
   @Output() ready = new EventEmitter<TObject>();
}

@Component()
export class NgtMesh extends NgtInstance<THREE.Mesh> {}
```

Here's a sample inheritance example. In reality, the inheritance level might be more than 1 level deep. You can see here that I can have generics with inheritance, and `(ready)` is strongly typed to `THREE.Mesh`

Let's try with Directive Composition

```ts
@Directive({
   selector: '[ngt-instance]'
   standalone: true
})
export class NgtInstance<TObject> {
   @Output() ready = new EventEmitter<TObject>();
}

@Component({
   hostDirectives: [
      // not sure how to provide THREE.Mesh here for (ready) to be strongly typed
      // maybe something like NgtInstance as NgtInstance<THREE.Mesh> and the compiler can figure it out?
      { directive: NgtInstance, outputs: ['ready'] }
   ]
})
export class NgtMesh {}
```

**Comment by crisbeto** [comment]
Regarding structural directives: I'm not sure why they need to be called out specifically. You can still apply a host directive to them, but it's going to work on the template instance, the same as if you were to apply a directive to the `ng-template`.

Regarding generics: this is one of the limitations we are aware of, but I'm not sure if there's a good way around it given that the host directives are referenced by value. We could introduce some way of inferring the generic type in the future, but it's not something we've considered yet for v1.

Regarding the standalone question: only the directives referenced in the `hostDirectives` array have to be standalone, the host itself can be imported into an `NgModule`, or it can `standalone` itself. In your example, `DirectiveA` and `DirectiveB` are standalone so they wouldn't need to be imported anywhere.

**Comment by nartc** [comment]
@crisbeto I just realized that we can probably have `selector: ng-template[component-selector]` and structural directive would work.

Thanks for the concise answers.

**Comment by Harpush** [comment]
@crisbeto Regarding genrics - won't Instantiation expressions solve it somehow?

**Comment by crisbeto** [comment]
@Harpush do you mean something like `hostDirectives: [new HostDir<T>()]`? I don't think that'll work. One possibility is to use a cast (e.g. `hostDirectives: [HostDir as HostDir<T>]`, but that might be weird as well.

**Comment by Harpush** [comment]
@crisbeto I meant the new typescript 4.7 feature allowing to assign generic params without `new`.
For example:
```ts
const Component = <T>() => ({hostDirectives: [Generic<T>]});
```
That is assuming you somehow have access to that `T` which seems like the real problem here (decorator can't access class generic params afaik).
But considering `T` is known at compile time this should work I think?
```ts
hostDirectives: [HostDir<number>]
```

**Comment by framasev** [comment]
Have you considered using a `@HostDirective` decorator and augmenting a directive through decorators composition?

```typescript
@HostDirective(FooDirective)
@Component()
class FooComponent {}

@HostDirective(FooDirective)
@Directive()
class BarDirective {}

@HostDirective(DirectiveA)
@HostDirective(DirectiveB)
@Directive()
class DirectiveC {}
```
